File: packages/java/src/main/java/io/github/qorechain/evm/EvmWindow.java
package io.github.qorechain.evm;

import io.github.qorechain.address.Address;
import io.github.qorechain.messages.QorechainMessages;
import io.github.qorechain.messages.TypedMessage;
import java.math.BigInteger;
import qorechain.pqc.v1.Tx;

/**
 * The EVM authorisation window (chain v3.2.0): message composers with the
 * chain's {@code ValidateBasic} bounds checked locally.
 *
 * <p>From chain v3.2.0 an EVM transaction is admitted only from an account that
 * has <strong>both</strong> a registered post-quantum key <strong>and</strong> an
 * open, unexhausted window. Opening one is an ordinary Cosmos-lane protobuf
 * message that travels the normal hybrid signing path — the classical key alone
 * can never open a window — so an EVM wallet keeps working unmodified
 * <em>inside</em> a window.
 *
 * <p>Three traps, all of which a wallet gets wrong by default:
 *
 * <ol>
 *   <li><strong>Ordering.</strong> QoreChain unifies the identity, so the Cosmos
 *       sequence <em>is</em> the EVM nonce, and opening a window advances it
 *       (measured: nonce 2 before, 3 after). Open the window, THEN read the
 *       nonce, THEN sign the EVM transaction; the other order gives "nonce too
 *       low".
 *   <li><strong>{@code maxValue} bounds value plus fees.</strong> It counts the
 *       transferred value AND the maximum fee each admitted transaction could
 *       pay (gas limit × gas fee cap), because the holder of the classical key
 *       sets the gas price. Measured on the testnet: a 1,000 uqor transfer with
 *       a 21,000 gas limit at 112.5 gwei consumed 3,363 uqor of the window.
 *       wei→uqor rounds UP.
 *   <li><strong>Never print "about 24 hours" for 17280 blocks.</strong> The chain
 *       constant says so at 5s blocks, but no QoreChain network runs at 5s: the
 *       testnet is at ~1.03 s (≈ 5 hours), mainnet at ~3.1 s (≈ 15 hours). Say
 *       "up to 17280 blocks", or compute from the chain's recent block time.
 * </ol>
 *
 * <p>There is deliberately <strong>no automatic opening</strong> here: a window
 * is an explicit authorisation the user makes.
 */
public final class EvmWindow {

    private EvmWindow() {}

    /**
     * The largest lifetime a window may be opened for ({@code MaxEVMWindowBlocks}).
     * See trap 3: do not present it to a user as "24 hours".
     */
    public static final long MAX_BLOCKS = 17280L;

    /**
     * The largest number of EVM transactions a window may admit
     * ({@code MaxEVMWindowTxs}). Zero is refused: a window that admits nothing is
     * a mistake, not a policy.
     */
    public static final long MAX_TXS = 1000L;

    /** The REST route for one account's window. */
    public static String restPath(String address) {
        return "/qorechain/pqc/v1/evm_window/" + address;
    }

    /**
     * Build a validated {@code MsgOpenEVMWindow}. {@code maxValueUqor} is an
     * integer amount of uqor as a decimal string (a {@code cosmos.Int} on the
     * wire, so it never passes through a float).
     *
     * <p>Every field is required: the chain refuses a missing one rather than
     * defaulting it. Opening replaces any existing window.
     *
     * @throws IllegalArgumentException if any bound is broken; the message names
     *     the bound.
     */
    public static TypedMessage open(String sender, long blocks, long maxTxs, String maxValueUqor) {
        return QorechainMessages.pqc.openEvmWindow(openMessage(sender, blocks, maxTxs, maxValueUqor));
    }

    /** The same as {@link #open} , returning the bare protobuf message. */
    public static Tx.MsgOpenEVMWindow openMessage(
            String sender, long blocks, long maxTxs, String maxValueUqor) {
        Tx.MsgOpenEVMWindow msg =
                Tx.MsgOpenEVMWindow.newBuilder()
                        .setSender(sender == null ? "" : sender)
                        .setBlocks(blocks)
                        .setMaxTxs(maxTxs)
                        .setMaxValue(maxValueUqor == null ? "" : maxValueUqor.trim())
                        .build();
        validateOpen(msg);
        return msg;
    }

    /**
     * Build a validated {@code MsgCloseEVMWindow}. Closing takes effect in the
     * same block.
     *
     * @throws IllegalArgumentException if the sender is not a {@code qor} address.
     */
    public static TypedMessage close(String sender) {
        return QorechainMessages.pqc.closeEvmWindow(closeMessage(sender));
    }

    /** The same as {@link #close} , returning the bare protobuf message. */
    public static Tx.MsgCloseEVMWindow closeMessage(String sender) {
        Tx.MsgCloseEVMWindow msg =
                Tx.MsgCloseEVMWindow.newBuilder().setSender(sender == null ? "" : sender).build();
        validateClose(msg);
        return msg;
    }

    /**
     * Mirror the chain's {@code ValidateBasic} for {@code MsgOpenEVMWindow} so a
     * caller fails fast, locally, instead of paying for a refused transaction:
     * sender is a valid {@code qor} bech32 address, {@code blocks} is in
     * 1..{@value #MAX_BLOCKS}, {@code max_txs} is in 1..{@value #MAX_TXS}, and
     * {@code max_value} is a positive integer amount of uqor.
     */
    public static void validateOpen(Tx.MsgOpenEVMWindow msg) {
        if (msg == null) {
            throw new IllegalArgumentException("open evm window: message is null");
        }
        validateSender(msg.getSender());
        long blocks = msg.getBlocks();
        if (blocks < 1 || blocks > MAX_BLOCKS) {
            throw new IllegalArgumentException(
                    "open evm window: blocks must be between 1 and "
                            + MAX_BLOCKS
                            + " (MaxEVMWindowBlocks), got "
                            + blocks);
        }
        long maxTxs = msg.getMaxTxs();
        if (maxTxs < 1 || maxTxs > MAX_TXS) {
            throw new IllegalArgumentException(
                    "open evm window: max_txs must be between 1 and "
                            + MAX_TXS
                            + " (MaxEVMWindowTxs), got "
                            + maxTxs);
        }
        String raw = msg.getMaxValue();
        if (raw == null || raw.isEmpty()) {
            throw new IllegalArgumentException(
                    "open evm window: max_value is required (an integer amount of uqor greater than 0)");
        }
        BigInteger value;
        try {
            value = new BigInteger(raw);
        } catch (NumberFormatException e) {
            throw new IllegalArgumentException(
                    "open evm window: max_value \"" + raw + "\" is not an integer amount of uqor");
        }
        if (value.signum() <= 0) {
            throw new IllegalArgumentException(
                    "open evm window: max_value must be greater than 0 uqor, got " + value);
        }
    }

    /** Mirror the chain's {@code ValidateBasic} for {@code MsgCloseEVMWindow}. */
    public static void validateClose(Tx.MsgCloseEVMWindow msg) {
        if (msg == null) {
            throw new IllegalArgumentException("close evm window: message is null");
        }
        validateSender(msg.getSender());
    }

    private static void validateSender(String sender) {
        if (sender == null || sender.isEmpty()) {
            throw new IllegalArgumentException("evm window: sender is required");
        }
        if (!Address.isValidBech32(sender, "qor")) {
            throw new IllegalArgumentException(
                    "evm window: sender \"" + sender + "\" is not a valid qor bech32 address");
        }
    }
}

File: packages/go/qorechain/messages/evm_window.go
package messages

import (
	"fmt"

	"cosmossdk.io/math"

	"github.com/qorechain/qorechain-sdk/packages/go/qorechain/address"
	pqcv1 "github.com/qorechain/qorechain-sdk/packages/go/qorechain/proto/qorechain/pqc/v1"
)

// Bounds of the EVM authorisation window, mirroring the chain's x/pqc
// ValidateBasic (chain v3.2.0). Every field of MsgOpenEVMWindow is required:
// the chain refuses a missing field rather than defaulting it.
const (
	// MaxEVMWindowBlocks is the largest lifetime a window may be opened for.
	//
	// The chain constant is documented as "about 24 hours at 5s blocks", but no
	// QoreChain network runs at 5s: the testnet produces a block about every
	// 1.03s (17280 blocks ≈ 5 hours) and mainnet about every 3.1s (≈ 15 hours).
	// Do NOT present this bound to a user as "24 hours" — say "up to 17280
	// blocks", or compute the duration from the chain's recent block time.
	MaxEVMWindowBlocks uint64 = 17280
	// MaxEVMWindowTxs is the largest number of EVM transactions a window may
	// admit. Zero is refused: a window that admits nothing is a mistake, not a
	// policy.
	MaxEVMWindowTxs uint64 = 1000
)

// ValidateOpenEVMWindow mirrors the chain's ValidateBasic for MsgOpenEVMWindow
// so a caller fails fast, locally, with a message naming the bound it broke
// instead of paying for a transaction the chain refuses.
//
// Checked: sender is a valid `qor` bech32 address; blocks is in 1..17280;
// max_txs is in 1..1000; max_value is set and strictly positive.
func ValidateOpenEVMWindow(msg *pqcv1.MsgOpenEVMWindow) error {
	if msg == nil {
		return fmt.Errorf("open evm window: message is nil")
	}
	if err := validateWindowSender(msg.Sender); err != nil {
		return err
	}
	if msg.Blocks == 0 || msg.Blocks > MaxEVMWindowBlocks {
		return fmt.Errorf("open evm window: blocks must be between 1 and %d (MaxEVMWindowBlocks), got %d", MaxEVMWindowBlocks, msg.Blocks)
	}
	if msg.MaxTxs == 0 || msg.MaxTxs > MaxEVMWindowTxs {
		return fmt.Errorf("open evm window: max_txs must be between 1 and %d (MaxEVMWindowTxs), got %d", MaxEVMWindowTxs, msg.MaxTxs)
	}
	if msg.MaxValue.IsNil() {
		return fmt.Errorf("open evm window: max_value is required (an integer amount of uqor greater than 0)")
	}
	if !msg.MaxValue.IsPositive() {
		return fmt.Errorf("open evm window: max_value must be greater than 0 uqor, got %s", msg.MaxValue.String())
	}
	return nil
}

// ValidateCloseEVMWindow mirrors the chain's ValidateBasic for
// MsgCloseEVMWindow: the sender must be a valid `qor` bech32 address.
func ValidateCloseEVMWindow(msg *pqcv1.MsgCloseEVMWindow) error {
	if msg == nil {
		return fmt.Errorf("close evm window: message is nil")
	}
	if err := validateWindowSender(msg.Sender); err != nil {
		return err
	}
	return nil
}

func validateWindowSender(sender string) error {
	if sender == "" {
		return fmt.Errorf("evm window: sender is required")
	}
	if !address.IsValidBech32(sender, "qor") {
		return fmt.Errorf("evm window: sender %q is not a valid qor bech32 address", sender)
	}
	return nil
}

// NewOpenEVMWindow builds a validated MsgOpenEVMWindow. maxValue is an integer
// amount of uqor as a decimal string (it is a cosmos.Int on the wire, so it is
// never parsed through a float).
//
// It bounds the transferred value PLUS the maximum fee each admitted
// transaction could pay (gas limit × gas fee cap), because the holder of the
// classical key sets the gas price: measured on the testnet, a 1,000 uqor
// transfer with a 21,000 gas limit at 112.5 gwei consumed 3,363 uqor of the
// window. wei→uqor rounds UP.
//
// Opening a window REPLACES any window the account already has, and — because
// QoreChain unifies the identity — it advances the account sequence, which is
// also the EVM nonce. Open the window, THEN read the nonce, THEN sign the EVM
// transaction; the other order gives "nonce too low".
func NewOpenEVMWindow(sender string, blocks, maxTxs uint64, maxValue string) (*pqcv1.MsgOpenEVMWindow, error) {
	if maxValue == "" {
		return nil, fmt.Errorf("open evm window: max_value is required (an integer amount of uqor greater than 0)")
	}
	value, ok := math.NewIntFromString(maxValue)
	if !ok {
		return nil, fmt.Errorf("open evm window: max_value %q is not an integer amount of uqor", maxValue)
	}
	msg := Pqc.OpenEVMWindow(sender, blocks, maxTxs, value)
	if err := ValidateOpenEVMWindow(msg); err != nil {
		return nil, err
	}
	return msg, nil
}

// NewCloseEVMWindow builds a validated MsgCloseEVMWindow. Closing takes effect
// in the same block.
func NewCloseEVMWindow(sender string) (*pqcv1.MsgCloseEVMWindow, error) {
	msg := Pqc.CloseEVMWindow(sender)
	if err := ValidateCloseEVMWindow(msg); err != nil {
		return nil, err
	}
	return msg, nil
}
